{"cells":[{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["from utils.FilePreparator import ADN_sequence_1, ADN_sequence_2, ARN_sequence_1, ARN_sequence_2\n","import Chatbot"]},{"cell_type":"markdown","metadata":{},"source":["# Start here\n","Evolution is the story of change over time, and at the heart of it all is DNA. Just as a typo in a book can change its meaning, mutations in DNA can change the proteins it codes for, leading to different traits in living organisms. Today, we'll step into the shoes of evolutionary biologists and explore how even small changes in DNA sequences can lead to dramatic variations in the appearance of the affected organism. Get ready to dive deep into the code of life and uncover the mechanisms behind genetic evolution!\n","# Biological Background\n","## The zebra fish\n","Zebre poisson ...  \n","\n","## Biological translation\n","One way to distinguish the two zebra fish gene sequences is to translate them into their corresponding polypeptides. So let's first recap how the translation is done in theory.  \n","\n","To translate a DNA sequence into a polypeptide (chain of amino acids), several steps are required. First, we need to translate the DNA sequence into the mRNA, which is complementary and transcribed. We can break this procedure down into two substeps:  \n","\n","<span style=\"color:orange\">Step 1.1:</span>  \n","The bases must be replaced by their counterparts (this sequence corresponds to the sequence of the opposite strand):  \n","| Brin 1 | Brin 2 |\n","|----------|----------|\n","| A    | T   |\n","| T    | A   |\n","| G    | C   |\n","| C    | G   |  \n","\n","<span style=\"color:magenta\">Step 1.2:</span>  \n","Then the bases thymine $T$ must be replaced by uracil $U$:\n","| Brin 2 | mRNA |\n","|----------|----------|\n","| A    | A   |\n","| _T_    | _U_   |\n","| G    | G   |\n","| C    | C   |  \n","\n","Now we have translated the DNA sequence in the corresponding mRNA sequence.  \n","\n","<span style=\"color:cyan\">Step 2:</span>  \n","Once we have found the mRNA sequence, we can use the genetic code to translate the sequence into the polypeptide.  \n","\n","## Goals\n","Since the translation would take far too long by hand, we want to write a program code with the programming language \"Python\" that does the work for us. You have never programmed before? No problem, just follow this template and you will learn:\n","1. intuitivly how to translate genetic code by hand.\n","2. interactively how to transfer your procedure to code.\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["# Stage 1\n","In the first Stage our goal is to translate a short sequence of DNA by hand and transfer our procedure into program code.  \n","## Tasks\n","__1a\\)__  \n","Translate the DNA sequence ``ADN_brin_1 = ATGAGGACA`` into the sequence of the opposite strand ``ADN_brin_2`` according to <span style=\"color:orange\">step 1.1:</span> described above.  \n","Take notes of how you proceeded.  \n","\n","__1b\\)__  \n","Write a Python program code that does the same as you did in __Task 1__. When you need help, open the ``Robofriend.ipynb`` file in a separate tab and start a conversation with the chatbot.  \n","It has no background information, so tell it what you have done so far and ask if it can help you with your current task.  \n","\n","__1c\\)__  \n","If your code works, you should now have produced a sequence called ``ADN_brin_2`` with code. Does it match your result from __Task 1__?  \n","If yes, you can proceed to the next stage, else try to find the error yourself or ask your teacher for help."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["run Chatbot"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Définis la séquence d'ADN que tu veux traduire.\n","ADN_sequence = \"AGTC\"#ADN_sequence_1"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Définir le dictionnaire du second brin de l'ADN.\n","ADN_switch_strand_dict = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Fonction de changement de brin d'ADN.\n","def switch_strand(ADN_sequence):\n","    ADN_switch_strand = \"\"\n","    for nucleotide in ADN_sequence:\n","        ADN_switch_strand += ADN_switch_strand_dict[nucleotide]\n","    return ADN_switch_strand\n","ADN_inverse = switch_strand(ADN_sequence)\n","print(ADN_inverse)"]},{"cell_type":"markdown","metadata":{},"source":["# Stage 2 \n","In this stage we will complete the program code to fully translate the DNA sequence into the according polypeptide.  \n","## Tasks\n","__2a\\)__  \n","execute the translation <span style=\"color:magenta\">step 1.2</span> to get a ``mRNA`` sequence and <span style=\"color:cyan\">step 2</span> to get a ``polypeptide`` sequence by hand. \n","\n","\n","__2b)__  \n","Write a Python program code that takes the ``ADN_brin_2`` produces a ``mRNA`` sequence by executing <span style=\"color:cyan\">step 2</span>. Use your previous results from __stage 1__ as a guide and compare the resulting sequence with the one you got in __Task 2a__.  \n","\n","__2c)__\n","Investigate the ``translate_genetic_sequence`` function and take notes. Do you understand what it does? If not, just ask the Chatbot.  \n","Then run the ``translate_genetic_sequence`` function using the ``mRNA`` sequence to produce a ``polypeptide`` sequence. Does it match your previous results."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Définir un dictionnaire pour la conversion ADN vers ARNm.\n","ADN_to_ARNm_dict = {'A': 'U', 'T': 'A', 'C': 'G', 'G': 'C'}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Fonction de conversion de l'ADN en ARNm.\n","def ADN_to_ARNm(sequence):\n","    ARNm_sequence = ''\n","    for nucleotide in sequence:\n","        ARNm_sequence += ADN_to_ARNm_dict[nucleotide]\n","    return ARNm_sequence\n","\n","ARNm_sequence = ADN_to_ARNm(ADN_inverse)\n","print(f\"ARNm sequence: {ARNm_sequence}\")"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Définir un dictionnaire pour les codons et les acides aminés correspondants.\n","codon_table = {\n","    \"UUU\": \"Phe\", \"UUC\": \"Phe\", \"UUA\": \"Leu\", \"UUG\": \"Leu\",\n","    \"CUU\": \"Leu\", \"CUC\": \"Leu\", \"CUA\": \"Leu\", \"CUG\": \"Leu\",\n","    \"AUU\": \"Ile\", \"AUC\": \"Ile\", \"AUA\": \"Ile\", \"AUG\": \"Met\",\n","    \"GUU\": \"Val\", \"GUC\": \"Val\", \"GUA\": \"Val\", \"GUG\": \"Val\",\n","    \"UCU\": \"Ser\", \"UCC\": \"Ser\", \"UCA\": \"Ser\", \"UCG\": \"Ser\",\n","    \"CCU\": \"Pro\", \"CCC\": \"Pro\", \"CCA\": \"Pro\", \"CCG\": \"Pro\",\n","    \"ACU\": \"Thr\", \"ACC\": \"Thr\", \"ACA\": \"Thr\", \"ACG\": \"Thr\",\n","    \"GCU\": \"Ala\", \"GCC\": \"Ala\", \"GCA\": \"Ala\", \"GCG\": \"Ala\",\n","    \"UAU\": \"Tyr\", \"UAC\": \"Tyr\", \"UAA\": \"FIN\", \"UAG\": \"FIN\",\n","    \"CAU\": \"His\", \"CAC\": \"His\", \"CAA\": \"Gln\", \"CAG\": \"Gln\",\n","    \"AAU\": \"Asn\", \"AAC\": \"Asn\", \"AAA\": \"Lys\", \"AAG\": \"Lys\",\n","    \"GAU\": \"Asp\", \"GAC\": \"Asp\", \"GAA\": \"Glu\", \"GAG\": \"Glu\",\n","    \"UGU\": \"Cys\", \"UGC\": \"Cys\", \"UGA\": \"FIN\", \"UGG\": \"Trp\",\n","    \"CGU\": \"Arg\", \"CGC\": \"Arg\", \"CGA\": \"Arg\", \"CGG\": \"Arg\",\n","    \"AGU\": \"Ser\", \"AGC\": \"Ser\", \"AGA\": \"Arg\", \"AGG\": \"Arg\",\n","    \"GGU\": \"Gly\", \"GGC\": \"Gly\", \"GGA\": \"Gly\", \"GGG\": \"Gly\",\n","}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Function to translate a genetic sequence into a protein\n","def translate_genetic_sequence(sequence):\n","    polypeptide = \"\"\n","    for i in range(0, len(sequence), 3):\n","        codon = sequence[i:i+3]\n","        amino_acid = codon_table.get(codon, \"-\")\n","        if amino_acid == \"FIN\":\n","            break\n","        else:\n","            polypeptide += amino_acid\n","\n","    return polypeptide\n"]},{"cell_type":"markdown","metadata":{},"source":["# Stage 3\n","In this final stage we will find out which of the DNA sequences ``ADN_sequence_1`` and ``ADN_sequence_2`` is the original version and which is the mutated version of the zebra fish gene.\n","\n","## Tasks\n","__3a)__  \n","In the beginning we used the short DNA sequence ``ATGAGGACA``, now replace it with ``ADN_sequence_1`` and run the following cell. Note the length of the polypeptide sequence and repeat the same with ``ADN_sequence_2``.\n","\n","The length of the two polypeptide sequences should be different. That your results from Task 3a do not reflect that, is because the ``translate_genetic_sequence`` function does not finish the sequence, if the stop codons ``UAA``, ``UAG`` or ``UGA`` appear.  \n","\n","__3b)__  \n","Find out what the ``translate_genetic_sequence`` function appends to the ``polypeptide`` when it processes a stop codon. Therefore Investigate the codon table.  \n","Using this information, implement a If Else statement into the ``translate_genetic_sequence`` function so it stops, when a stop codon is processed. Ask the chatbot if you need help.  \n","\n","__3c)__  \n","Check again the length of ``ADN_sequence_1`` and ``ADN_sequence_2``. Can you tell now which sequence is the original and which is the mutated version?  \n","If yes, well done you are a programming genie and well prepared to study bio informatics."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Translate the genetic sequence into a protein\n","polypeptide = translate_genetic_sequence(ARNm_sequence)\n","\n","# Display the translated protein sequence\n","print(\"Translated polypeptide sequence:\", polypeptide)\n","print(\"Length of polypeptide sequence:\", int(len(polypeptide)/3))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.3"}},"nbformat":4,"nbformat_minor":2}
